src/DiagonalHessianApproximation.jl

```diff
@@ -47,22 +50,33 @@ positive definite.
 """
 function DiagonalQN(d::AbstractVector{T}, psb::Bool = false) where {T <: Real}
   prod = (res, v, α, β) -> mulSquareOpDiagonal!(res, d, v, α, β)
-  DiagonalQN(d, length(d), length(d), true, true, prod, prod, prod, 0, 0, 0, true, true, true, psb)
+  n = length(d)
+  s = similar(d)
+  s2 = similar(d)
+  y = similar(d)
+  DiagonalQN(d, s, s2, y, n, n, true, true, prod, prod, prod, 0, 0, 0, true, true, true, psb)
 end
 
 # update function
 # s = x_{k+1} - x_k
 # y = ∇f(x_{k+1}) - ∇f(x_k)
 function push!(
   B::DiagonalQN{T, I, V, F},
-  s::V,
-  y::V,
+  s0::V,
+  y0::V,
 ) where {T <: Real, I <: Integer, V <: AbstractVector{T}, F}
-  s2 = (si^2 for si ∈ s)
-  trA2 = dot(s2, s2)
-  if trA2 == 0
-    error("Cannot divide by zero and trA2 = 0")
+  s = B.s
+  s2 = B.s2
+  y = B.y
+  s0Norm = norm(s0, 2)
+  if s0Norm == 0
+    error("Cannot update DiagonalQN operator with s=0")
   end
+  # sᵀBs = sᵀy can be scaled by ||s||² without changing the updating problem
+  s .= s0 ./ s0Norm
+  y .= y0 ./ s0Norm
+  s2 .= s .^ 2
+  trA2 = dot(s2, s2)
```

Comment:
Instead of allocating vectors, couldn't we create iterators
```julia
s2 = (si^2 / s0Norm for si ∈ s)
y = (yi / s0Norm for yi ∈ y0) 
```

Reply:
That's allocating no?

```julia
julia> s = rand(1000);
julia> @allocated (si^2 / 2.0 for si ∈ s)
133703
```

Reply:
It should be just 16 bytes if you put it in a function, but I wish we could make it zero.